lib/helpers/common.js

```diff
@@ -67,7 +67,19 @@ function flatten(update, path, options, schema) {
  * ignore
  */
 
-function modifiedPaths(update, path, result) {
+function modifiedPaths(update, path, result, recursion=null) {
+  if(recursion == null){
+    recursion = {
+      count: 0,
+      raw: {update, path}
+    };
+  }
+
+  recursion.count++;
+  if(recursion.count >= 1024){
```

Comment:
I think this is a bit brittle. Admittedly we're unlikely to have depth >= 1024 unless there's an infinite recursion. But I think a better approach would be to use a `WeakSet` or `WeakMap` to track what objects we've seen in the recursion, similar to how [`getIndexes()` does it](https://github.com/Automattic/mongoose/blob/master/lib/helpers/schema/getIndexes.js#L16-L29).